Studio-Lovelies/GG-JointJustice-Unity
# Implement compile-time constant action mapping inside ActionDirector

## Summary
As talked about, this introduces an easier way to add or modify actions to the ActionDirector.

Adding an action literally requires one new method now. ([example](https://github.com/Studio-Lovelies/GG-JointJustice-Unity/blob/b02171c25ef967864e99cb62a52f20f26874cd64/Assets/Scripts/TextDecoder/ActionDecoder.cs#L325-L335))
[All existing methods have been ported.](https://github.com/Studio-Lovelies/GG-JointJustice-Unity/blob/b02171c25ef967864e99cb62a52f20f26874cd64/Assets/Scripts/TextDecoder/ActionDecoder.cs#L103-L341)

Just ensure that...
  - The method is **private**
  - The method's **name only consists of capital letter and underscores** (`[A-Z_]+`) and matches what you want the Ink script to use
  - **Optional parameters** are added in **regular C#-syntax** (i.e. `private void FOO(string bar = "default")`
  - All **parameters** have a **parser for their type**. [If unused types are used, appropriate error messages are thrown, once the decoder encounters such a line.](https://github.com/Studio-Lovelies/GG-JointJustice-Unity/blob/b02171c25ef967864e99cb62a52f20f26874cd64/Assets/Scripts/TextDecoder/ActionDecoder.cs#L62-L77)  
    The currently supported types are:
    - [bool](https://github.com/Studio-Lovelies/GG-JointJustice-Unity/blob/b02171c25ef967864e99cb62a52f20f26874cd64/Assets/Scripts/TextDecoder/Parser/BoolParser.cs)
    - [float](https://github.com/Studio-Lovelies/GG-JointJustice-Unity/blob/b02171c25ef967864e99cb62a52f20f26874cd64/Assets/Scripts/TextDecoder/Parser/FloatParser.cs)
    - [int](https://github.com/Studio-Lovelies/GG-JointJustice-Unity/blob/b02171c25ef967864e99cb62a52f20f26874cd64/Assets/Scripts/TextDecoder/Parser/IntParser.cs)
    - [string](https://github.com/Studio-Lovelies/GG-JointJustice-Unity/blob/b02171c25ef967864e99cb62a52f20f26874cd64/Assets/Scripts/TextDecoder/Parser/StringParser.cs)
    - [ItemDisplayPosition enums](https://github.com/Studio-Lovelies/GG-JointJustice-Unity/blob/b02171c25ef967864e99cb62a52f20f26874cd64/Assets/Scripts/TextDecoder/Parser/ItemDisplayPositionParser.cs)  

---

This change also introduces **auto-generated tests** for parsing **each action**. If a method is added and respects the guidelines above, the following tests are automatically generated for it:
- A **valid** line for any given action
  - For actions with optional parameters, **all possible cases are run** (without, with some and with all optional parameters present)
- A line with **too few** arguments for any given action
  - Actions that support **no parameters** at all, are exempt
- A line with **too many** arguments for any given action
- A line with the correct number, but **incorrect types of arguments** for any given action
  - Actions that **only accept `string`s** are exempt

Tests can be found in the `ActionDecoderTests` category.  
To see the actual Inky line, that's being tested, run the test, click on it and check the line under `Attempting to parse:`:
![zSydHHwryk](https://user-images.githubusercontent.com/1689033/139543144-33ba071d-858b-4509-ae33-248c9b3e1041.gif)

## Testing instructions
1. Switch to this branch
2. Play the game (I've run the current `Inky-TestScene` with the `Ross_Cool_X` narrative script
3. Ensure all `PlayMode`- and `EditMode`-tests pass

## Additional information

- `[ ]` Changes UI
- `[ ]` Introduces new feature
- `[ ]` Removes existing feature
- `[ ]` Has associated resource:
